src/crimson/common/local_shared_foreign_ptr.h

```diff
@@ -34,7 +34,9 @@ class local_shared_foreign_ptr {
 
   /// Wraps a pointer object and remembers the current core.
   local_shared_foreign_ptr(seastar::foreign_ptr<PtrType> &&fptr)
-    : ptr(seastar::make_lw_shared(std::move(fptr))) {}
+    : ptr(seastar::make_lw_shared(std::move(fptr))) {
+    assert(ptr && *ptr);
```

Comment:
reset() suggests that the constraint should be  ```assert(!ptr || (ptr && *ptr));```

Reply:
Can `fptr` be nullable in `local_shared_foreign_ptr`?

Looks get_foreign() and get_owner_shard() suggests that `assert(ptr && *ptr)`, that if there is a foreign_ptr, it should not be null.

Reply:
I think my original intent was for it not to be nullable, but the public reset() implementation added below would allow users to create null instances.  Is that the intent?

Reply:
Yeah, my understanding is that the `fptr` (foreign_ptr<PtrType>) isn't nullable, and `local_shared_foreign_ptr::ptr` is nullptr only after `reset()`.

Also, it's user's responsibility not to use local_shared_foreign_ptr after `reset()` until it is assigned another non-null `fptr`.

It's also reasonable to me to relax the above constraint, which implies to replace all the `assert(ptr && *ptr)` by `assert(!ptr || (ptr && *ptr));` and to handle the `!ptr` case properly.

What do you think?

Reply:
A public method (reset() in this case) that puts the ptr into a null state means that the ptr is nullable.  Operations that require it to be not null should still assert ```assert(ptr && *ptr);``` but comparisons, assignments, and move/copy constructors should work for any valid instance, including null.

Reply:
See the updated https://github.com/ceph/ceph/pull/50835/commits/f96ae0b367825c0e4433a57a086e03f75b1eab2e, I'm leaving operator*() and operator->() to be not null.